protostar/commands/test/environments/fuzz_test_execution_environment.py

```diff
@@ -72,72 +86,56 @@ async def invoke(self, function_name: str) -> TestExecutionResult:
             parameters
         ), f"{self.__class__.__name__} expects at least one function parameter."
 
+        strategy_selector = StrategySelector(parameters)
+
         self.set_cheatcodes(
             FuzzTestCaseCheatcodeFactory(
                 state=self.state,
                 expect_revert_context=self._expect_revert_context,
                 finish_hook=self._finish_hook,
+                strategy_selector=strategy_selector,
             )
         )
 
         self.set_custom_hint_locals(
-            [TestContextHintLocal(self.state.context), CairoStructHintLocal()]
+            [
+                TestContextHintLocal(self.state.context),
+                CairoStructHintLocal(),
+                StrategiesHintLocal(),
+            ]
         )
 
         execution_resources: List[ExecutionResourcesSummary] = []
 
         database = InMemoryExampleDatabase()
-        strategy_selector = StrategySelector(parameters)
-
-        runs_counter = RunsCounter()
+        runs_counter = RunsCounter(budget=self._fuzz_config.max_examples)
 
         # NOTE: Hypothesis' ``reporter`` global is a thread local variable.
         #   Because we are running Hypothesis from separate thread, and the test itself is
         #   running in a separate thread executor, we must set the ``reporter`` each first time
         #   we invoke Hypothesis code in new thread.
 
-        @seed(TestingSeed.current())
-        @settings(
-            database=database,
-            deadline=None,
-            print_blob=False,
-            report_multiple_bugs=False,
-            verbosity=HYPOTHESIS_VERBOSITY,
-            max_examples=self._fuzz_config.max_examples,
-        )
-        @given(data_object=data())
-        async def test(data_object: DataObject):
-            self.fork_state_for_test()
-
-            run_no = next(runs_counter)
-            with self.state.output_recorder.redirect(("test", run_no)):
-                with with_reporter(protostar_reporter):
-                    inputs: Dict[str, Any] = {}
-                    for param in strategy_selector.parameter_names:
-                        search_strategy = strategy_selector.search_strategies[param]
-                        inputs[param] = data_object.draw(search_strategy, label=param)
-
+        def test_thread():
+            with with_reporter(protostar_reporter):
+                for _ in range(self._fuzz_config.max_strategy_learnings):
```

Comment:
Can you add clarification comment why rebuilding test function is necessary for each learning?